dhctl/pkg/operations/mirror/chunk_writer.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2023 Flant JSC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mirror
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"path/filepath"
+)
+
+type chunkedFileWriter struct {
+	chunkSize  int64
+	chunkIndex int
+
+	workingDir   string
+	baseFileName string
+	activeChunk  *os.File
+}
+
+func newChunkWriter(chunkSize int64, dirPath, baseFileName string) *chunkedFileWriter {
+	return &chunkedFileWriter{
+		chunkSize:    chunkSize,
+		workingDir:   filepath.Clean(dirPath),
+		baseFileName: baseFileName,
+	}
+}
+
+func (c *chunkedFileWriter) Write(p []byte) (int, error) {
+	if len(p) == 0 {
+		return 0, nil
+	}
+
+	if c.activeChunk == nil {
+		if err := c.swapActiveChunk(); err != nil {
+			return 0, fmt.Errorf("Create first chunk: %w", err)
+		}
+	}
+
+	chunkStat, err := c.activeChunk.Stat()
+	if err != nil {
+		return 0, fmt.Errorf("Read chunk size: %w", err)
+	}
+
+	buf := bytes.NewBuffer(p)
+	bytesWritten := 0
+	for {
+		for c.chunkSize-chunkStat.Size() > 0 {
+			s := buf.Next(512 * 1024)
+			if len(s) == 0 {
+				return bytesWritten, nil
+			}
+
+			written, err := c.activeChunk.Write(s)
+			bytesWritten += written
+			if err != nil {
+				return 0, fmt.Errorf("Write to chunk: %w", err)
+			}
+
+			chunkStat, err = c.activeChunk.Stat()
+			if err != nil {
+				return 0, fmt.Errorf("Read chunk size: %w", err)
+			}
+		}
+
+		if err = c.swapActiveChunk(); err != nil {
+			return 0, fmt.Errorf("Swap active chunk: %w", err)
+		}
+		chunkStat, err = c.activeChunk.Stat()
+		if err != nil {
+			return 0, fmt.Errorf("Read chunk size: %w", err)
+		}
+	}
+}
+
+func (c *chunkedFileWriter) Close() error {
+	if c.activeChunk != nil {
+		if err := c.activeChunk.Sync(); err != nil {
```

Comment:
duplicate code with https://github.com/deckhouse/deckhouse/pull/8000/files#diff-99dfcd465f16919e5a94f704f592f20ff1f866574994d8e4c018d61a8829b49eR102

We can create method `closeActiveChunk` for it